relayer/src/link/relay_path.rs

```diff
@@ -1278,27 +1279,109 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
-    /// to the target chain.
+    /// Drives the relaying of elapsed operational data items meant for
+    /// a specified target chain forward.
     ///
-    /// This method performs relaying using the asynchronous sender.
-    /// Retains the operational data as pending, and associates it
-    /// with one or more transaction hash(es).
-    pub fn execute_schedule(&self) -> Result<(), LinkError> {
-        let (src_ods, dst_ods) = self.try_fetch_scheduled_operational_data()?;
+    /// Given an iterator of `OperationalData` elements, this function
+    /// first determines whether the current piece of operational data
+    /// has elapsed.
+    ///
+    /// A piece of operational data is considered 'elapsed' if it has surpassed
+    /// its target chain's:
+    /// 1. Latest timestamp
+    /// 2. Maximum block time
+    /// 3. Latest height
+    ///
+    /// If the current piece of operational data has elapsed, then relaying
+    /// is performed using the asynchronous sender. Operational data is
+    /// retained as pending and is associated with one or more transaction
+    /// hash(es).
+    ///
+    /// Should an error occur when attempting to relay a piece of operational
+    /// data, this function returns all subsequent unprocessed pieces of
+    /// operational data back to the caller so that they can be re-queued
+    /// for processing; the operational data that failed to send is dropped.
+    ///
+    /// Note that pieces of operational data that have not elapsed yet are
+    /// also placed in the 'unprocessed' bucket.
+    fn execute_schedule_for_target_chain<I: Iterator<Item = OperationalData>>(
+        &mut self,
+        mut operations: I,
+        target_chain: OperationalDataTarget,
+    ) -> Result<VecDeque<OperationalData>, (VecDeque<OperationalData>, LinkError)> {
+        let mut unprocessed = VecDeque::new();
+
+        while let Some(od) = operations.next() {
+            let elapsed_result = match target_chain {
+                OperationalDataTarget::Source => od.has_conn_delay_elapsed(
+                    &|| self.src_time_latest(),
+                    &|| self.src_max_block_time(),
+                    &|| self.src_latest_height(),
+                ),
+                OperationalDataTarget::Destination => od.has_conn_delay_elapsed(
+                    &|| self.dst_time_latest(),
+                    &|| self.dst_max_block_time(),
+                    &|| self.dst_latest_height(),
+                ),
+            };
 
-        for od in dst_ods {
-            let reply =
-                self.relay_from_operational_data::<relay_sender::AsyncSender>(od.clone())?;
+            match elapsed_result {
+                Ok(elapsed) => {
+                    if elapsed {
+                        match self
+                            .relay_from_operational_data::<relay_sender::AsyncSender>(od.clone())
+                        {
+                            Ok(reply) => self.enqueue_pending_tx(reply, od),
```

Comment:
The logic here is quite deep (~5 lvls of indentation) and I'm having difficulties understanding it. I appreciate the thorough doc comments for the method, which help, and wondering if we should have a helper inner `fn` here or some other approach to make things more obvious.

Reply:
I couldn't figure out a clean way to encapsulate this logic in a separate function. In some cases this function would need to return a single piece of operational data (if it hasn't elapsed), in other cases it needs to return `(VecDeque<OperationalData>, Option<Error>)`; I don't see a clean way to structure the return type. 

I've opted instead to annotate the nested logic as clearly as possible. Hopefully that addresses your concern?